Calibration/HcalCalibAlgos/plugins/HcalHBHEMuonAnalyzer.cc

```diff
@@ -281,265 +287,290 @@ void HcalHBHEMuonAnalyzer::analyze(const edm::Event& iEvent, const edm::EventSet
 	
-	MuonEcalDetId.push_back((trackID.detIdECAL)()); 
-	MuonHcalDetId.push_back((trackID.detIdHCAL)());  
-	MuonEHcalDetId.push_back((trackID.detIdEHCAL)());
+	ecalDetId_.push_back((trackID.detIdECAL)()); 
+	hcalDetId_.push_back((trackID.detIdHCAL)());  
+	ehcalDetId_.push_back((trackID.detIdEHCAL)());
   
 	HcalDetId check;
 	std::pair<bool,HcalDetId> info = spr::propagateHCALBack(pTrack,  geo, bField, (((verbosity_/100)%10>0)));
 	if (info.first) { 
 	  check = info.second;
 	}	
 
-	if (trackID.okECAL) {
+	bool okE = trackID.okECAL;
+	if (okE) {
 	  const DetId isoCell(trackID.detIdECAL);
 	  std::pair<double,bool> e3x3 = spr::eECALmatrix(isoCell,barrelRecHitsHandle,endcapRecHitsHandle,*theEcalChStatus,geo,caloTopology,sevlv.product(),1,1,-100.0,-100.0,-500.0,500.0,false);
 	  eEcal = e3x3.first;
+	  okE   = e3x3.second;
 	}
+#ifdef EDM_ML_DEBUG
+	edm::LogInfo("HBHEMuon") << "Propagate Track to ECAL: " << okE << ":"
+				 << trackID.okECAL << " E "
+				 << eEcal << std::endl;
+#endif
 
 	if (trackID.okHCAL) {
 	  const DetId closestCell(trackID.detIdHCAL);
 	  HcalDetId hcidt(closestCell.rawId());  
-	  if ((hcidt.ieta() == check.ieta()) && (hcidt.iphi() == check.iphi())) 
+	  if ((hcidt.ieta() == check.ieta()) && (hcidt.iphi() == check.iphi()))
 	    tmpmatch= true;
 
+	  HcalSubdetector subdet = HcalDetId(closestCell).subdet();
+	  int             ieta   = HcalDetId(closestCell).ieta();
+	  int             iphi   = HcalDetId(closestCell).iphi();
+	  bool            hborhe = (std::abs(ieta) == 16);
+
 	  eHcal = spr::eHCALmatrix(theHBHETopology, closestCell, hbhe,0,0, false, true, -100.0, -100.0, -100.0, -100.0, -500.,500.,useRaw_);
 	  std::vector<std::pair<double,int> > ehdepth;
 	  spr::energyHCALCell((HcalDetId) closestCell, hbhe, ehdepth, maxDepth_, -100.0, -100.0, -100.0, -100.0, -500.0, 500.0, useRaw_, (((verbosity_/1000)%10)>0));
 	  for (unsigned int i=0; i<ehdepth.size(); ++i) {
 	    eHcalDepth[ehdepth[i].second-1] = ehdepth[i].first;
+	    HcalSubdetector subdet0 = (hborhe) ? ((ehdepth[i].second >= depthHE) ? HcalEndcap : HcalBarrel) : subdet;
+	    HcalDetId hcid0(subdet0,ieta,iphi,ehdepth[i].second);
+	    double actL = activeLength(DetId(hcid0));
+	    activeL[ehdepth[i].second-1] = actL;
+	    activeLengthTot += actL;
+#ifdef EDM_ML_DEBUG
+	    if ((verbosity_%10) > 0)
+	      edm::LogInfo("HBHEMuon") << hcid0 << " E " << ehdepth[i].first
+				       << " L " << actL << std::endl;
+#endif
 	  }
 	  
-	  HcalDetId hcid0(closestCell.rawId());
-	  activeL = activeLength(trackID.detIdHCAL);
 	  HcalDetId           hotCell;
 	  spr::eHCALmatrix(geo, theHBHETopology, closestCell, hbhe, 1,1, hotCell, false, useRaw_, false);
 	  isHot = matchId(closestCell,hotCell);
 	  if (hotCell != HcalDetId()) {
+	    subdet = HcalDetId(hotCell).subdet();
+	    ieta   = HcalDetId(hotCell).ieta();
+	    iphi   = HcalDetId(hotCell).iphi();
+	    hborhe = (std::abs(ieta) == 16);
```

Comment:
hi @bsunanda - this line seems wrong? 
otherwise, could you avoid the 3x creation of the HcalDetId just above?